refactor(parser): Avoid peek in parse_delimited_list

leaysgur opened the pull request:
Out of curiosity, I've just deleted the peeking in `parse_delimited_list`.

Then, coverage errors show me there are only a few places missing error report for invalid trailing comma after rest element, etc.

So I added those checks manually.

In TSC, these checks seem to be done at checker, not parser. But we need this in parser for JS.

---

(Sorry, I have PASSED the test, but I am not very confident that this change is intrinsically sound.)

leaysgur commented:
At this point, there appears to be no regression in the parser benchmark.

But strangely, when I added a commit that removes the **unused** `Lexer::lookahead` related things, the regression occurs!

...🤔

Boshen commented:
### Merge activity

* **May 28, 8:30 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **May 28, 8:30 AM UTC**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

leaysgur committed:
refactor(parser): Avoid peek in parse_delimited_list (#11343)

Out of curiosity, I've just deleted the peeking in `parse_delimited_list`.

Then, coverage errors show me there are only a few places missing error report for invalid trailing comma after rest element, etc.

So I added those checks manually.

In TSC, these checks seem to be done at checker, not parser. But we need this in parser for JS.

---

(Sorry, I have PASSED the test, but I am not very confident that this change is intrinsically sound.)
